src/Features/Core/Portable/ForeachToFor/AbstractForeachToForCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,346 @@
+        {
+            return type == null || type is IErrorTypeSymbol;
+        }
+
+        private static IMethodSymbol GetInterfaceMembers(ITypeSymbol interfaceType, string memberName)
+        {
+            var members = interfaceType.GetMembers(memberName);
+            if (!members.IsEmpty)
+            {
+                return (IMethodSymbol)members[0];
+            }
+
+            foreach (var current in interfaceType.AllInterfaces)
+            {
+                var member = GetInterfaceMembers(current, memberName);
+                if (member != null)
+                {
+                    return member;
+                }
+            }
+
+            return null;
+        }
+
+        private static bool CheckRequireCollectionStatement(IOperation operation)
+        {
+            return operation.Kind != OperationKind.LocalReference && operation.Kind != OperationKind.FieldReference;
+        }
+
+        private IOperation RemoveImplicitConversion(IOperation collection)
+        {
+            if (collection is IConversionOperation conversion && conversion.IsImplicit)
+            {
+                return RemoveImplicitConversion(conversion.Operand);
+            }
+
+            return collection;
+        }
+
+        private bool CheckForeachVariable(SemanticModel semanticModel, ISymbol foreachVariable, SyntaxNode foreachStatement)
+        {
+            var (start, end) = GetForeachBody(foreachStatement);
+            var dataFlow = semanticModel.AnalyzeDataFlow(start, end);
+
+            if (!dataFlow.Succeeded)
+            {
+                // if we can't get good analysis, assume it is written
+                return false;
+            }
+
+            return !dataFlow.WrittenInside.Contains(foreachVariable);
+        }
+
+        private async Task<Document> ConvertForeachToForAsync(
+            Document document,
+            ForeachInfo foreachInfo,
+            CancellationToken cancellationToken)
+        {
+            var model = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+            var workspace = document.Project.Solution.Workspace;
+            var editor = new SyntaxEditor(model.SyntaxTree.GetRoot(cancellationToken), workspace);
+
+            ConvertToForStatement(model, foreachInfo, editor, cancellationToken);
+
+            var newRoot = editor.GetChangedRoot();
+            return document.WithSyntaxRoot(newRoot);
+        }
+
+        protected class ForeachInfo
+        {
+            public ForeachInfo(
+                string countName, string explicitCastInterface, bool requireCollectionStatement, SyntaxNode foreachStatement)
+            {
+                CountName = countName;
+
+                ExplicitCastInterface = explicitCastInterface;
+                RequireExplicitCast = explicitCastInterface != null;
+                RequireCollectionStatement = requireCollectionStatement || RequireExplicitCast;
```

Comment:
storing this as data unnecessary.

Reply:
sure, I can make the property to calculate it. but why that matter? this is not a state that will held in memory for long time.

Reply:
State is usually harder to reason about than computation.  When you have this state then someone has to understand "what are the values this can be, and how do they relate to the other values".  For example, i have to think "can RequireExplicitCast be true, but RequireCollectionStatement be false"?  

With computation the question is answered because the property says what it means.